### DISCUSSION TITLE
DXF to DXF conversion from other versions.

### DISCUSSION THREAD

**Discussion by Bruno-Rafael-Almeida** [mozman/ezdxf#904] [Q&A]
Good afternoon everyone, I'm having to develop a way to convert .dxf files from one version to another (in this case I'm having to convert from a more recent version to an older one, so that it works on older machines).
But I couldn't find anything like that in the documentation, I'm studying a little about the structure of .dxf files so that I understand better, but I would like to know if there is any way to do this using this library, the ezdxf for python.
Thanks!

**Comment by mozman** [comment]
`Ezdxf` is not a DXF converter, because DXF is not well documented and I don't wanna provide support for a feature that can fail for a million reasons.

The best choice is the free [ODA File Converter](https://www.opendesign.com/guestfiles/oda_file_converter) for Windows/Linux/macOS. Copy the [odafc](https://github.com/mozman/ezdxf/blob/master/src/ezdxf/addons/odafc.py) add-on into your project if you just need the Python integration of the converter.

`Ezdxf` can do some conversions reliable if the file contains only basic graphics like LINE, ARC, CIRCLE.

The following tips can work but are not recommended/reliable for complex DXF files.

Upgrading from DXF R2000+ to a newer version often work by assigning the new DXF version:

```Python
import ezdxf

doc = ezdxf.readfile("r2000.dxf")
doc.dxfversion = "R2013"
doc.saveas("R2013.dxf")
```

Downgrading of DXF versions R2004-2018 to R2000 and newer can work for files with simple content because `ezdxf` does not export unsupported entities and attributes. This simple downgrading does not work for DXF R12. The `r12export` add-on (new in v1.1) can export any DXF version as DXF R12. This  add-on approximates newer entities like SPLINE and ELLIPSE as polylines, converts LWPOLYLINE and MESH entities to their older counterpart POLYLINE and explodes MTEXT into simple TEXT entities, unsupported DXF attributes (line weight, true color) are lost.

To all beginners: downgrading/upgrading of your 500+ MB "AutoCAD Map/Civil/Architect" file does not work!
